### PR TITLE
20250801 reference entries

### LINKED ISSUE #3014: Reference Types / Sub Entries
We have a need for certain classes of objects to be stored in "referenced" entries rather than directly in the parent.

Examples include:

* Images related to oauth2/domain
* Sessions and Descendant Session state
* Scope Maps / Claim Maps in OAuth2

Now lets explain why - currently these types are stored directly in the entry via valuesets. This creates a few problems.

* Images consume memory, and so caching of the entry in ram implies that the image must be in ram, even if not actively being read.
* To read the image, necessitates loading *all* of the entry, not just the image.
* To add new types we need to continually extend ValueSet and DBEntry which is non-trivial and presents complexity to contributors rather than having a way to create compound-sub-entries.

For example, sessions as subentries would allow re-use of existing uuid, datetime and simpler types in a sub-entry, rather than requiring a more complex valueset item to be made. It means that when we get the entry we can choose if we want to load sessions or not. 

In the image case it also allows *deduplication* of identical images so that on upload if two items reference identical content, they can then share the same image reference.

### PR DESCRIPTION
**Pull request by Firstyear**
# Change summary

- Implement reference entries 

Fixes #3014 

Checklist

- [x] This PR contains no AI generated code
- [ ] book chapter included (if relevant)
- [ ] design document included (if relevant)

### CODE REVIEW

**Comment by Firstyear** [comment]
FYI still not ready for review, I need to add more tests.

**Comment by Firstyear** [comment]
Includes prevention of reference loops, and replication conflict handling. I might think of some edge cases in future, but I'm pretty sure this is solid rn. 

There are some other schema cleanups in here that I found in the process.

File: server/lib/src/server/recycle.rs [hunk 1]
```diff
@@ -203,6 +208,29 @@ impl QueryServerWriteTransaction<'_> {
 
         let norm_cand: Vec<Entry<_, _>> = res?;
 
+        // Finally, setup the mod for restoring memberships from direct member of
+        let mut dm_mods: BTreeMap<Uuid, ModifyList<ModifyInvalid>> = Default::default();
+
+        for entry in &pre_candidates {
+            // Get this entries uuid.
```
Comment (yaleman) [inline server/lib/src/server/recycle.rs]:
```suggestion
            // Get this entry's uuid.
```

File: server/lib/src/plugins/refint.rs [hunk 2]
```diff
@@ -43,9 +40,8 @@ impl ReferentialIntegrity {
         // will fail. This will return the union of the inclusion after the
         // operation.
         let filt_in = filter!(f_inc(inner));
-        let b = qs.internal_exists(filt_in).map_err(|e| {
-            admin_error!(err = ?e, "internal exists failure");
-            e
+        let b = qs.internal_exists(&filt_in).inspect_err(|err| {
+            error!(?err, "internal exists failure");
```
Comment (yaleman) [inline server/lib/src/plugins/refint.rs]: Something like this to show what failed, maybe? "internal exists failure" isn't actionable for an error and the same message is here and a lower message, so that wouldn't help for debugging.

```suggestion
            error!(filter=%filt_in, ?err, "internal exists failure");
```

File: server/lib/src/schema.rs [hunk 3]
```diff
@@ -777,6 +777,8 @@ impl SchemaWriteTransaction<'_> {
 
     #[instrument(level = "debug", name = "schema::generate_in_memory", skip_all)]
     pub fn generate_in_memory(&mut self) -> Result<(), OperationError> {
+        //
+        // THIS IS FOR SYSTEM CRITICAL INTERNAL SCHEMA ONLY
```
Comment (yaleman) [inline server/lib/src/schema.rs]: This should be a function docstring 😆

File: server/lib/src/schema.rs [hunk 4]
```diff
@@ -1493,6 +1495,41 @@ impl SchemaWriteTransaction<'_> {
                 syntax: SyntaxType::ReferenceUuid,
             },
         );
+
+        self.attributes.insert(
+            Attribute::Refers,
+            SchemaAttribute {
+                name: Attribute::Refers,
+                uuid: UUID_SCHEMA_ATTR_REFERS,
+                description: String::from("A reference to linked object"),
```
Comment (yaleman) [inline server/lib/src/schema.rs]:
```suggestion
                description: String::from("A reference to another object"),
```

File: server/lib/src/schema.rs [hunk 5]
```diff
@@ -1493,6 +1495,41 @@ impl SchemaWriteTransaction<'_> {
...
+                uuid: UUID_SCHEMA_ATTR_CASCADE_DELETED,
+                description: String::from("A marker attribute denoting that this entry was deleted by cascade when this UUID was deleted."),
+                multivalue: false,
+                unique: false,
+                phantom: false,
+                sync_allowed: false,
+                replicated: Replicated::True,
+                indexed: true,
+                // NOTE: This has to be Uuid so that referential integrity doesn't consider
+                // this value in it's operation.
```
Comment (yaleman) [inline server/lib/src/schema.rs]:
```suggestion
                // this value in its operation.
```

File: server/lib/src/server/delete.rs [hunk 6]
```diff
@@ -49,6 +49,56 @@ impl QueryServerWriteTransaction<'_> {
...
+                .iter()
+                .map(|entry| { f_eq(Attribute::Refers, PartialValue::Refer(entry.get_uuid())) })
+                .collect(),
+        ));
+
+        let mut pre_cascade_delete_candidates = self
+            .internal_search(references_filt)
+            .inspect_err(|err| error!(?err, "unable to find reference entries"))?;
+
+        #[cfg(debug_assertions)]
```
Comment (yaleman) [inline server/lib/src/server/delete.rs]:
```suggestion
        #[cfg(any(test,debug_assertions))]
```

Just 'cause I'm paranoid

File: server/lib/src/plugins/refint.rs [hunk 7]
```diff
@@ -121,6 +121,77 @@ impl ReferentialIntegrity {
...
+            })
+            .collect();
+
+        // F_inc(lusion). All items of inner must be 1 or more, or the filter
+        // will fail. This will return the union of the inclusion after the
+        // operation.
+        let filt_in = filter!(f_inc(inner));
+
+        let b = qs.internal_exists(&filt_in).inspect_err(|err| {
+            error!(?err, "internal exists failure");
```
Comment (yaleman) [inline server/lib/src/plugins/refint.rs]: This could use an actionable message and at least a slightly different message for debugging

File: server/lib/src/plugins/refint.rs [hunk 8]
```diff
@@ -121,6 +121,77 @@ impl ReferentialIntegrity {
...
+        }
+
+        let mut looping = Vec::with_capacity(inner.len());
+        for u in inner {
+            let filt_in = filter!(f_and(vec![
+                f_eq(Attribute::Uuid, PartialValue::Uuid(*u)),
+                f_pres(Attribute::Refers)
+            ]));
+            let b = qs.internal_exists(&filt_in).inspect_err(|err| {
+                error!(?err, "internal exists failure");
```
Comment (yaleman) [inline server/lib/src/plugins/refint.rs]: See above 😆

File: server/lib/src/plugins/refint.rs [hunk 9]
```diff
@@ -179,6 +250,60 @@ impl Plugin for ReferentialIntegrity {
...
+            .iter()
+            .copied()
+            .map(|conflict_uuid| f_eq(Attribute::Refers, PartialValue::Refer(conflict_uuid)))
+            .collect::<Vec<_>>();
+
+        let filt = filter!(f_or(filt_inner));
+
+        // Does anything match?
+        let conflict_cand = qs.internal_exists(&filt).inspect_err(|err| {
+            error!(?err, "internal exists error");
```
Comment (yaleman) [inline server/lib/src/plugins/refint.rs:277]: 😏

File: server/lib/src/plugins/refint.rs [hunk 10]
```diff
@@ -415,36 +547,57 @@ impl ReferentialIntegrity {
...
+            Self::check_refers_to_target_loop_fast(qs, refers_target_uuids.as_slice())?;
+
+        if loops_present {
+            let looping_uuids =
+                Self::check_refers_to_target_loop_slow(qs, refers_target_uuids.as_slice())?;
+
+            error!("some entries introduce an invalid reference to another reference.");
+            for looping in looping_uuids {
+                error!(?looping);
+            }
```
Comment (yaleman) [inline server/lib/src/plugins/refint.rs]:
```suggestion
            error!(looping_uuids = ?looping_uuids,
            "entries introduce an invalid reference to another reference.");
```
Just to cut the noise a bit due to how verbose tracing tree is
